crates/trie/sparse-parallel/src/lower.rs

```diff
@@ -0,0 +1,97 @@
+}
+
+impl Default for LowerSparseSubtrie {
+    /// Creates a new blinded subtrie with no allocated storage.
+    fn default() -> Self {
+        Self::Blind(None)
+    }
+}
+
+impl LowerSparseSubtrie {
+    /// Returns a reference to the underlying [`SparseSubtrie`] if this subtrie is revealed.
+    ///
+    /// Returns `None` if the subtrie is blinded (has no nodes).
+    pub(crate) fn as_revealed_ref(&self) -> Option<&SparseSubtrie> {
+        match self {
+            Self::Blind(_) => None,
+            Self::Revealed(subtrie) => Some(subtrie.as_ref()),
+        }
+    }
+
+    /// Returns a mutable reference to the underlying [`SparseSubtrie`] if this subtrie is revealed.
+    ///
+    /// Returns `None` if the subtrie is blinded (has no nodes).
+    pub(crate) fn as_revealed_mut(&mut self) -> Option<&mut SparseSubtrie> {
+        match self {
+            Self::Blind(_) => None,
+            Self::Revealed(subtrie) => Some(subtrie.as_mut()),
+        }
+    }
+
+    /// Reveals the lower [`SparseSubtrie`], transitioning it from the Blinded to the Revealed
+    /// variant, preserving allocations if possible.
+    ///
+    /// The given path is the path of a node which will be set into the [`SparseSubtrie`]'s `nodes`
+    /// map immediately upon being revealed. If the subtrie is blinded, or if its current root path
+    /// is shorter than this one, than this one becomes the new root path of the subtrie.
+    pub(crate) fn reveal(&mut self, path: &Nibbles) {
+        match self {
+            Self::Blind(None) => {
+                *self = Self::Revealed(Box::new(SparseSubtrie::new(*path)));
+            }
+            entry @ Self::Blind(Some(_)) => {
+                let Self::Blind(Some(mut subtrie)) = core::mem::take(entry) else { unreachable!() };
+                subtrie.path = *path;
+                *entry = Self::Revealed(subtrie);
+            }
+            Self::Revealed(subtrie) => {
+                if path.len() < subtrie.path.len() {
+                    subtrie.path = *path;
+                }
+            }
+        };
+    }
+
+    /// Clears the subtrie and transitions it to the blinded state, preserving a cleared
+    /// [`SparseSubtrie`] if possible.
+    pub(crate) fn clear(&mut self) {
+        *self = match core::mem::take(self) {
+            Self::Blind(allocated) => Self::Blind(allocated),
+            Self::Revealed(mut subtrie) => {
+                subtrie.clear();
+                Self::Blind(Some(subtrie))
+            }
+        }
+    }
+
+    /// Takes ownership of the underlying [`SparseSubtrie`] if revealed and the predicate returns
+    /// true.
+    ///
+    /// If the subtrie is revealed, and the predicate function returns `true` when called with it,
+    /// then this method will take ownership of the subtrie and transition this `LowerSparseSubtrie`
+    /// to the blinded state. Otherwise, returns `None`.
+    pub(crate) fn take_if<P>(&mut self, predicate: P) -> Option<Box<SparseSubtrie>>
```

Comment:
I don't see it being used, do we need it?

Reply:
It's used in `take_changed_lower_subtries`: https://github.com/paradigmxyz/reth/blob/242c64d22e82272e494f829b3deecdd73d422afa/crates/trie/sparse-parallel/src/trie.rs#L974-L978

Definitely open to better ideas of how to handle this case though, I wasn't super happy with this